content/blog/2019-10-28-self/index.md

```diff
@@ -0,0 +1,176 @@
++++
+title = "An Efficient Implementation of Self"
+
+[extra]
+bio = """
+  [Rachit Nigam](https://rachitnigam.com) is a second year PhD student interested in
+  programming languages & computer architecture. In his free time, he
+  [subtweets](https://twitter.com/notypes/status/1170037148290080771) his advisor and [annoys tenured professors](https://twitter.com/natefoster/status/1074401015565291520).
+"""
+[[extra.authors]]
+name = "Rachit Nigam"
+link = "https://rachitnigam.com"
++++
+
+[An Efficient Implementation of Self, a Dynamically-Typed Object-Oriented
+Language Based on Prototypes][paper] presents techniques for runtime
+compilation, now more commonly referred to as Just In Time (JIT) compilation,
+for the [Self][] programming language.
+
+> We have developed and implemented techniques that double the performance of
+> the dynamically-typed object oriented languages.
+
+
+## The Challenge
+
+The term "dynamic language" is mostly commonly associated with modern scripting
+languages like Python and JavaScript. Self, a much older language developed
+at the famed Xerox PARC labs takes the philosophy of dynamism to a logical
+extreme -- *everything* in self is a message to an object. This includes
+Java-like method calls on object *as well as* control structures like loops
+and conditionals.
+
+For example, the Self condition `IfTrue:IfFalse:` is a method invocation on
+the boolean object `true` and `false`. This means that a simple conditional:
+
+```
+if (x) 1 else 2
+```
+
+in Self parlance (written with JavaScript-esque syntax) looks like this:
+
+
+```
+x.If({ true: () => 1, false: () => 2 })
+```
+
+The runtime behavior of this program is invoking the method `IfTrue` on the
+object `x`, which can be a `true` or `false` but is not required to be those,
+and execute the "thunk" (a function with no argument) corresponding to the true
+or the false branch.  Note that `x` is not required to be `true` or `false`.
+Any Self object can define the `If` method and specify its conditional
+behavior.
```

Comment:
This paragraph switches a few times between several names for the same method: `If`, `IfTrue`, and `IfTrue:IfFalse:`. Maybe pick one?